decoder.go

```diff
@@ -0,0 +1,211 @@
+package redis
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// Decoder represents a type with a custom Decode() function that can
+// take a string and parse it to set a desired value on a struct field,
+// similar to how encoding/json implements UnmarshalJSON().
+type Decoder interface {
+	Decode(string) error
+}
+
+// defDecoder represents decoding functions for default built-in types.
+type defDecoder func(reflect.Value, string) error
+
+// structField represents a single field in a target struct.
+type structField struct {
+	index int
+
+	// Decoder.Decode() or defDecoder()
+	fn interface{}
+}
+
+var (
+	defDecoders = map[reflect.Kind]defDecoder{
```

Comment:
defDecoders -> decoders
map[reflect.Kind]defDecoder -> []defDecoder

Also these types are usually called standard / buit-in / primitives, not default.